File: Python/day03.py
#!/usr/bin/env python
# coding: utf-8

import re
import os

txtFileName = input("Enter the name of the .txt file to read (suggestion: input_day03.txt): ").strip()
if not os.path.exists(txtFileName):
    print(f"Error: File '{txtFileName}' not found.")
    exit(1)

with open(txtFileName, "r") as file:
    text = file.read()

mul = r"mul\((\d{1,3}),(\d{1,3})\)"

part = input("Which part do you want to run? (1 or 2): ").strip()

if part == "1":
    # Part 1
    matches = re.findall(mul, text)
    resultMul = sum(int(x) * int(y) for x, y in matches)
    print("The sum of all the multiplications is: {}".format(resultMul))
elif part == "2":
    # Part 2
    enabled = True
    resultMul = 0
    sections = re.split(r"(do\(\)|don\'t\(\))", text)
    
    for section in sections:
        if 'do()' in section:
            enabled = True
        elif 'don\'t()' in section:
            enabled = False

        for match in re.findall(mul,section):
            x,y = match
            if enabled:
                resultMul += int(x)*int(y)
    print("The sum of all the multiplications is: {}".format(resultMul))
else:
    print("Invalid selection. Please enter 1 or 2.")

input("\nPress Enter to exit.")

File: cpp/day03.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <regex>
#include <filesystem>

int main() {
    std::string txtFileName;
    std::cout << "Enter the name of the .txt file to read (suggestion: input_day03.txt): ";
    std::getline(std::cin, txtFileName);

    if (!std::filesystem::exists(txtFileName)) {
        std::cerr << "Error: File '" << txtFileName << "' not found.\n";
        return 1;
    }

    std::ifstream file(txtFileName);
    std::string text((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
    file.close();

    std::regex mul_pattern(R"(mul\((\d{1,3}),(\d{1,3})\))");
    std::regex do_dont_pattern(R"(do\(\)|don't\(\))");

    std::string part;
    std::cout << "Which part do you want to run? (1 or 2): ";
    std::getline(std::cin, part);

    if (part == "1") {
        // Part 1
        int resultMul = 0;
        std::sregex_iterator it(text.begin(), text.end(), mul_pattern);
        std::sregex_iterator end;
        for (; it != end; ++it) {
            int x = std::stoi((*it)[1]);
            int y = std::stoi((*it)[2]);
            resultMul += x * y;
        }
        std::cout << "The sum of all the multiplications is: " << resultMul << "\n";
    }
    else if (part == "2") {
        // Part 2
        int resultMul = 0;
        bool enabled = true;

        std::sregex_token_iterator token_it(text.begin(), text.end(), do_dont_pattern, {-1, 0});
        std::sregex_token_iterator end;
        for (; token_it != end; ++token_it) {
            std::string segment = token_it->str();
            if (segment == "do()") {
                enabled = true;
                continue;
            } else if (segment == "don't()") {
                enabled = false;
                continue;
            }

            if (enabled) {
                std::sregex_iterator match_it(segment.begin(), segment.end(), mul_pattern);
                std::sregex_iterator match_end;
                for (; match_it != match_end; ++match_it) {
                    int x = std::stoi((*match_it)[1]);
                    int y = std::stoi((*match_it)[2]);
                    resultMul += x * y;
                }
            }
        }
        std::cout << "The sum of all the multiplications is: " << resultMul << "\n";
    }
    else {
        std::cout << "Invalid selection. Please enter 1 or 2.\n";
    }

    std::cout << "\nPress Enter to exit.";
    std::cin.get();
    return 0;
}
